packages/esm-amd-loader/src/esm-amd-loader.ts

```diff
@@ -18,42 +18,349 @@ interface Window {
+}
+interface BaseWaiting {
+  /**
+   * The dependencies of this module, in order.
+   */
+  deps: Module[];
+  /**
+   * Args that we will pass into to moduleBody.
+   */
+  args: Array<{}>;
+  /**
+   * The body of the module, which is executed after its dependencies are
+   * loaded.
+   *
+   * In AMD/Commonjs terminology, this is the factory function.
+   */
+  moduleBody: undefined|Function;
+}
+interface WaitingOnEarlierScripts extends BaseWaiting {
+  state: StateEnum.WaitingOnEarlierScripts;
+}
+interface WaitingOnDeps extends BaseWaiting {
+  state: StateEnum.WaitingOnDeps;
+}
+interface Executed {
+  state: StateEnum.Executed;
+}
+interface Failed {
+  state: StateEnum.Failed;
+  readonly error: Error;
+}
+type ModuleStateData =
+    Initialized|Loading|WaitingOnEarlierScripts|WaitingOnDeps|Executed|Failed;
+
+interface Module<SD extends ModuleStateData = ModuleStateData> {
   url: NormalizedUrl;
   urlBase: NormalizedUrl;
   exports: {[id: string]: {}};
-  /** This is a top-level module. */
+  stateData: SD;
+  /** True if this is a top-level module. */
   isTopLevel: boolean;
-  /** This module hasn't started loading its script yet. */
-  needsLoad: boolean;
-  /** All dependencies are resolved and the factory has run. */
-  resolved: boolean;
-  error: Error|undefined;
-  /** Callbacks from dependents waiting for this module to resolve. */
-  onResolve: Array<() => void>;
-  onError: Array<ErrorCallback>;
+  /**
+   * Callbacks that are called exactly once, for the next time the module
+   * progresses to a new state.
+   */
+  onNextStateChange: Array<() => void>;
 }
 
-type ResolveCallback = (...args: Array<{}>) => void;
-type ErrorCallback = (error: Error) => void;
-type NormalizedUrl = string&{_normalized: never};
-
-(function() {
-
 /**
  * A global map from a fully qualified module URLs to module objects.
  */
 const registry: {[url: string]: Module} = Object.create(null);
-
-let pendingDefine: ((mod: Module) => void)|undefined = undefined;
+let pendingDefine: (() => [Array<string>, ResolveCallback | undefined])|
+    undefined = undefined;
 let topLevelScriptIdx = 0;
 let previousTopLevelUrl: string|undefined = undefined;
 let baseUrl = getBaseUrl();
 
+/** Begin loading a module from the network. */
+function load(module: Module<Initialized>): Module<Loading> {
+  const mutatedModule = stateTransition(module, {state: StateEnum.Loading});
+
+  const script = document.createElement('script');
+  script.src = module.url;
+
+  script.onload = () => {
+    let deps: string[], moduleBody;
+    if (pendingDefine !== undefined) {
+      [deps, moduleBody] = pendingDefine();
+    } else {
+      // The script did not make a call to define(), otherwise the global
+      // callback would have been set. That's fine, we can resolve immediately
+      // because we don't have any dependencies, by definition.
+      deps = [];
+      moduleBody = undefined;
+    }
+    beginWaitingOnEarlierScripts(mutatedModule, deps, moduleBody);
```

Comment:
why do we get to start waiting on earlier scripts now? Seems like we should be waiting on deps first. Is that name really descriptive of what it's doing?